src/nbl/video/COpenGL_Queue.h

```diff
@@ -249,6 +258,12 @@ class COpenGL_Queue final : public IGPUQueue
             if (!IGPUQueue::submit(_count, _submits, _fence))
                 return false;
 
+            if (_fence)
+            {
+                COpenGLFence* glfence = static_cast<COpenGLFence*>(_fence);
+                glfence->setToBeSignaled();
```

Comment:
why are all the fences, semaphores and events not sharing the glsync object!?

Reply:
there should be one `COpenGLSync` created here, and it should be passed to the signal functions of the semaphores and queues which will be signalled.

Reply:
all semaphores share a GL sync. fence has separate one because fence signal is separate request to queue thread

Reply:
they should all share, there's really no reason not to

Reply:
> fence signal is separate request to queue thread

thats the reason

Reply:
why is it even a request?

the creation of a `COpenGLSync` should be the request (or even just the initialization, creation of the interal `GLSync` object), the `IOpenGLSyncPrimitiveBase::signal()` should just be a move of the externally created `core::smart_refctd_ptr<COpenGLSync>`

Reply:
> the creation of a COpenGLSync should be the request

hm not bad idea

Reply:
actually one step further, allow creation of `COpenGLSync` but request the initialization, that way you wont have to wait for the request to complete.

```cpp
        inline COpenGLSync(IOpenGL_LogicalDevice* _dev, IOpenGL_FunctionTable* _gl, bool lockToQueue=false) : device(_dev), lockedTable(lockToQueue ? _gl:nullptr), haveNotWaitedOnQueueMask(~(0x1ull<<_gl->getGUID())), cachedRetval(ES_TIMEOUT_EXPIRED), sync(nullptr) {}
        
        inline void initialize(IOpenGL_FunctionTable* _gl)
        {
            sync = _gl->glSync.pglFenceSync(GL_SYNC_GPU_COMMANDS_COMPLETE, 0);
            if (!lockedTable)
                _gl->glGeneral.pglFlush();
        }

        E_STATUS waitCPU(IOpenGL_FunctionTable* _gl, uint64_t timeout)
        {
            assert(!lockedTable || lockedTable==_gl);
            if (cachedRetval != ES_TIMEOUT_EXPIRED)
                return cachedRetval;
            else if (!sync) // has not been intiailized yet, so could not have been signalled yet
                return ES_TIMEOUT_EXPIRED;

            GLenum status = _gl->glSync.pglClientWaitSync(sync, lockedTable ? GL_SYNC_FLUSH_COMMANDS_BIT:0u, timeout);
            switch (status)
            {
            case GL_ALREADY_SIGNALED:
                return (cachedRetval = ES_ALREADY_SIGNALED);
                break;
            case GL_TIMEOUT_EXPIRED:
                return (cachedRetval = ES_TIMEOUT_EXPIRED);
                break;
            case GL_CONDITION_SATISFIED:;
                return (cachedRetval = ES_CONDITION_SATISFIED);
                break;
            default:
                break;
            }
            return (cachedRetval = ES_FAIL);
        }

        void waitGPU(IOpenGL_FunctionTable* _gl)
        {
            if (!lockedTable) // OpenGL device does not need to wait on itself within the same context
            {
                const uint64_t queueMask = 0x1ull<<_gl->getGUID();
                if (haveNotWaitedOnQueueMask&queueMask)
                {
                    while (!sync) {} // has not been initialized yet, so could not have been enqueued to be signalled yet
                    _gl->glSync.pglWaitSync(sync, 0, GL_TIMEOUT_IGNORED);
                    haveNotWaitedOnQueueMask ^ = queueMask;
                }
            }
            else
            {
                assert(lockedTable==_gl);
            }
        }

        inline GLsync getOpenGLName() const
        {
            return sync;
        }

    private:
        IOpenGL_LogicalDevice* device;
        IOpenGL_FunctionTable* lockedTable;
        uint64_t haveNotWaitedOnQueueMask;
        E_STATUS cachedRetval;
        GLsync sync; // TODO: make it volatile or atomic
```

Reply:
You can only lock the table if the `COpenGLSync` will be created for use as part of signalling a batch of `IGPUEvent`s all of which have the `VK_EVENT_CREATE_DEVICE_ONLY_BIT_KHR` property